ibis/common/caching.py

```diff
@@ -46,3 +56,67 @@ def __delitem__(self, key):
 
     def __repr__(self):
         return f"{self.__class__.__name__}({self._data})"
+
+
+class RefCountedCache:
+    """A cache with reference-counted keys.
+
+    We could implement `MutableMapping`, but the `__setitem__` implementation
+    doesn't make sense and the `len` and `__iter__` methods aren't used.
+
+    We can implement that interface if and when we need to.
+    """
+
+    def __init__(self, *, populate, lookup, finalize) -> None:
+        from ibis import util
+
+        self.cache = bidict()
+        self.refs = Counter()
+        self.populate = populate
+        self.lookup = lookup
+        self.finalize = finalize
+        self.names = defaultdict(partial(util.generate_unique_table_name, "cache"))
```

Comment:
Ideally generate unique table name should be passed as a callback too.

Reply:
Fixed.